tripy/tripy/frontend/trace/ops/matmul.py

```diff
@@ -16,87 +16,32 @@
 #
 
 from dataclasses import dataclass
+from typing import Dict, List
 
 import tripy.frontend.trace.ops.utils as op_utils
-from tripy import utils, constraints
-from tripy.frontend import utils as frontend_utils
+from tripy import constraints
+from tripy.common.exception import raise_error
 from tripy.frontend.ops.registry import TENSOR_METHOD_REGISTRY
 from tripy.frontend.trace.ops.base import BaseTraceOp
-from tripy.common.exception import raise_error
 
 
 @dataclass(repr=False)
 class MatrixMultiplication(BaseTraceOp):
+
+    contracting_dim: Dict[str, List[int]]
+    batching_dim: Dict[str, List[int]]
+
     def __str__(self):
         return f"{self.outputs[0].name} = {' @ '.join([inp.name for inp in self.inputs])}"
 
     def infer_rank(self):
+        out_rank = max(self.inputs[0].rank, self.inputs[1].rank)
         if self.inputs[0].rank == 1 and self.inputs[1].rank == 1:
-            self.outputs[0].rank = 0
-
-        assert len(self.inputs) == 2, f"MatrixMultiplication expects exactly two inputs."
-        a_rank = self.inputs[0].rank
-        b_rank = self.inputs[1].rank
-
-        for index, rank in enumerate([a_rank, b_rank]):
-            if rank < 1:
-                utils.raise_error_io_info(
-                    self,
-                    "Input tensors must have at least 1 dimension.",
-                    details=[
-                        f"Inputs for operation: '@' must have at least one dimension, but input {index} has rank: {rank}"
-                    ],
-                )
-
-        if a_rank == 1 and b_rank == 1:
-            # case 1: both operands are 1-D
-            self.batching_dim = {"lhs": [], "rhs": []}
-            self.contracting_dim = {"lhs": [0], "rhs": [0]}
-            self.outputs[0].rank = 0
-        else:
-            # stablehlo dot_general requires same number of batching dims for lhs, rhs.
-
-            def compute_contracting_dims(rank_a, rank_b):
-                if rank_a < 1 or rank_b < 1:
-                    raise ValueError("Ranks must be at least 1")
-
-                def is_vector(rank):
-                    return rank == 1
-
-                def is_matrix(rank):
-                    return rank >= 2
+            out_rank = 0
+        elif self.inputs[0].rank == 1 or self.inputs[1].rank == 1:
+            out_rank = 1
```

Comment:
should we add an assertion check for unsupported input ranks?

Reply:
Yes, there's already one [here](https://github.com/NVIDIA/TensorRT-Incubator/pull/328/files#diff-143063c187b705db2521def8d47c2e26f18d468dadae71de1ec02272b410efdfR186).

Reply:
Oh that's handled before the conditional:
```
out_rank = max(self.inputs[0].rank, self.inputs[1].rank)
```
I could move it to an else block if preferred.